lib/path/tests/prop.nix

```diff
@@ -57,4 +67,33 @@ let
       # Return an empty string when failed
       if tryOnce.success then tryOnce.value else "";
 
-in lib.genAttrs strings normaliseAndCheck
+  # Checks the subpathJoin properties for a list of path-like values
+  subpathJoinCheck = list:
+    let
+      allValid = all path.subpath.isValid list;
+
+      try = builtins.tryEval (path.subpath.join list);
+      normalisedConcat = path.subpath.normalise (concatStringsSep "/" list);
+    in
+      # It should only succeed when there's only valid subpaths passed
+      assert assertMsg
+        (allValid -> try.success)
+        "Even though all strings ${pretty list} are valid as a subpath, subpath.join failed";
+      assert assertMsg
+        (! allValid -> ! try.success)
+        "Even though some string in ${pretty list} is invalid as a subpath, subpath.join succeeded";
+
+      # Check the equivalence to the normalised concatStringsSep. All other
+      # properties can be proven from this one
+      assert assertMsg
+        (allValid && list != [] -> try.value == normalisedConcat)
+        "For the strings ${pretty list}, subpath.join returned \"${try.value}\" instead of the expected \"${normalisedConcat}\"";
+      true;
+
+  # Check the properties for all lists of strings
+  subpathJoinCheckAll =
+    assert all subpathJoinCheck lists;
+    null;
+
+in builtins.seq subpathJoinCheckAll
+  (genAttrs strings normaliseAndCheck)
```

Comment:
This is a bit ugly. Given that we will extend the test suite anyway (I haven't looked at the other PRs yet), maybe we should do a little bit of generalisation. I feel uneasy about the change to the randomly generated paths, because it makes the whole setup more complicated and significantly harder to understand before getting into the actual testing. Have you considered approaching this differently? Instead of producing lists of paths in the random generator, produce a sequence of random numbers (which can be reused for other purposes), pass that and do the chunking in Nix.

Maybe we could also have one Nix file per test, and import each from an entry point that does the pre-processing of the input? Then each file can take whatever data type it cares about, directly. The main file will then only combine the tests in a way that produces output as needed for further processing back in the shell. I suppose, with the other tests, we will have a series of asserts and then produce a value, but having that in one place without the other noise will allow for easier changes of factoring.

Reply:
It would be nice not to have the generation logic in Nix expressions, to decouple the Nix tests from the test runner. However, we also use this pattern for invoking external commands that we model in Nix, such as `realpath`, so we can't easily get away from all coupling between framework and tests.

I do think it's worth exploring for the sake of writing Nix-only tests in a consistent style where all domain-specific testing logic is in the same language.
Instead of generating examples or a large chunk of randomness ahead of time, we could easily create a pseudorandom number generator by invoking `builtins.hashString "sha256"` repeatedly on a state variable, starting from a small seed that's generated outside of Nix. This achieves the goal without ever running out of randomness, and therefore without any coupling with the non-Nix world. The state variable can be handled by a simplified Nix equivalent of the [quickcheck `Gen`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck-Gen.html) library. Instead of applicative we could have a single function that takes an attrset of "gens" and a function matching those attr names, etc.

Reply:
I don't agree we should push toward implementing more things in the Nix language, even if it's possible and not all that hard to do. It will happen anyway, but should happen much more slowly and deliberately. That is, I wouldn't oppose a QuickCheck implementation or PRs to migrate existing tests to it. But for the sake of testing this library, simple Nix expressions that batch-process inputs should do.

Reply:
Since this is test code, we could maybe even allow using the cursed `builtins.exec`:

```
$ nix-instantiate --eval -E 'builtins.exec [ "realpath" "-m" "//foo/./bar/." ]' --allow-unsafe-native-code-during-evaluation
/foo/bar
```

But no I don't think it would be a great idea to write all tests in Nix itself. Nix is just not made for this kind of thing. One of the main restrictions is that we can't check for all kinds of evaluation failures. `builtins.tryEval` only works for `throw` and `assert`, no other failures, and it doesn't allow checking for the error message either. Randomness too is not something Nix was made for. Yes we can hack around it with `builtins.hashString`, custom PRNG's (I even wrote one in Nix myself some time ago) and monadic generator passing, but that's gonna be painful.

I think something like https://github.com/Mic92/pythonix is the right way to go. Unfortunately that project is archived now, but it works by linking to `libexpr` directly and using that to evaluate expressions. This way failures can be caught and checked. The main problem with this is that `libexpr` isn't stable, so it requires continuous maintenance. But using that approach testing libraries of generic programming languages can be used.

The alternative of just using the CLI as a stable API has the problem that it's way too slow (this is why I opted for just a single CLI call in these tests here).

I'll see if I can do some more decoupling for the property tests here though. Maybe having a separate file/folder for every property to check and using the `.nix` file only for batch evaluation and no checking.

Reply:
I think it would be better to just drop the property tests from this PR. I don't want to have this be blocked on improved property tests. There's still unit tests as well.